Upkeep new actions & sources

feyzullah opened the pull request:
Added new actions and sources
Could not add `Work Order Category Change` because as I see, Work Order Category cannot be changed after Work Order is placed.
To QA Reviewer; `create-request` and `find-vendor` works only for enterprise plan. Also webhook registration cannot be done with the Business Plus trial plan. I double checked all the things I could not test, so they should be working but we need an Enterprise plan to test them.

<a href="https://gitpod.io/#https://github.com/PipedreamHQ/pipedream/pull/4160"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>

dylburger commented:
Thanks for submitting this PR! When we review PRs, we follow the [Pipedream component guidelines](https://pipedream.com/docs/components/guidelines/). If you're not familiar, here's a quick checklist:

- [ ] [Create components to address specific use cases whenever possible](https://pipedream.com/docs/components/guidelines/#component-scope)
- [ ] [Component `key`s should follow the format `app_name_slug-slugified-component-name`](https://pipedream.com/docs/components/guidelines/#component-key-pattern)
- [ ] Components should follow the [standard directory structure](https://pipedream.com/docs/components/guidelines/#folder-structure)
- [ ] [Prefer Node.js client libraries to REST APIs](https://pipedream.com/docs/components/guidelines/#using-apis-vs-client-libraries)
- [ ] When making API requests, [handle pagination](https://pipedream.com/docs/components/guidelines/#pagination) to ensure all data / events are processed
- [ ] [Use `secret` props](https://pipedream.com/docs/components/guidelines/#capturing-sensitive-data) to capture sensitive data
- [ ] Props and methods should be defined in [app files](https://pipedream.com/docs/components/guidelines/#app-files) whenever possible
- [ ] Document methods with [JS Docs](https://pipedream.com/docs/components/guidelines/#methods)
- [ ] [Use `optional` props](https://pipedream.com/docs/components/guidelines/#optional-vs-required-props) whenever possible, and [set a `default` value](https://pipedream.com/docs/components/guidelines/#default-values) where you can
- [ ] Use [async options](https://pipedream.com/docs/components/guidelines/#async-options) to accept user input wherever possible
- [ ] [Create subscription webhooks sources (vs polling sources) whenever possible](https://pipedream.com/docs/components/guidelines/#webhook-vs-polling-sources)
- [ ] [Always emit a `summary` with events](https://pipedream.com/docs/components/guidelines/#emit-a-summary)
- [ ] [Use a built-in deduping strategy wherever possible](https://pipedream.com/docs/components/guidelines/#deduping)
- [ ] When building webhook sources, [follow these best practices](https://pipedream.com/docs/components/guidelines/#webhook-sources)
- [ ] When building polling sources, [follow these best practices](https://pipedream.com/docs/components/guidelines/#polling-sources)

lcaresia commented on components/upkeep/actions/create-purchase-order/create-purchase-order.mjs:
```diff
@@ -0,0 +1,28 @@
+      $,
+      data: this.prepareData(),
+    });
+    $.export("$summary", `Purchase order with ID:${result.id} has been created.`);
```
```suggestion
    $.export("$summary", `Purchase order with ID ${result.id} has been created.`);
```

lcaresia commented on components/upkeep/actions/create-request/create-request.mjs:
```diff
@@ -0,0 +1,79 @@
+        dueDate: this.dueDate,
+      },
+    });
+    $.export("$summary", `Request with ID:${result.id} has been created.`);
```
```suggestion
    $.export("$summary", `Request with ID ${result.id} has been created.`);
```

lcaresia commented on components/upkeep/actions/create-work-order/create-work-order.mjs:
```diff
@@ -0,0 +1,128 @@
+        additionalUsers: this.additionalUsers,
+      },
+    });
+    $.export("$summary", `Request with ID:${result.id} has been created.`);
```
```suggestion
    $.export("$summary", `Request with ID ${result.id} has been created.`);
```

lcaresia commented on components/upkeep/actions/update-purchase-order/update-purchase-order.mjs:
```diff
@@ -0,0 +1,36 @@
+      purchaseOrderId: this.purchaseOrderId,
+      data: this.prepareData(),
+    });
+    $.export("$summary", `Purchase order with ID:${result.id} has been updated.`);
```
```suggestion
    $.export("$summary", `Purchase order with ID ${result.id} has been updated.`);
```

lcaresia commented on components/upkeep/sources/common/common.mjs:
```diff
@@ -0,0 +1,61 @@
+      throw new Error("getTime is not implemented!");
+    },
+  },
+  hooks: {
+    async activate() {
+      const { webhookId } = await this.app.createWebhook({
+        title: this.getTitle(),
+        endpoint: this.http.endpoint,
+        events: this.getEvents(),
+      });
+      this._setHookID(webhookId);
+      console.log(`Created webhook. (Hook ID: ${webhookId}, endpoint: ${this.http.endpoint})`);
+    },
+    async deactivate() {
+      await this.app.deleteWebhook({
+        webhookId: this._getHookID(),
+      });
+    },
+  },
```
Is possible implement webhook `deploy()` here?

lcaresia reviewed (changes requested):
Thanks for your contribution!

Just some fixes!

lcaresia commented on components/upkeep/upkeep.app.mjs:
```diff
@@ -1,11 +1,319 @@
+    },
+    _getHeaders(headers = {}) {
+      return {
+        "Session-Token": `${this.$auth.oauth_access_token}`,
```
Move this a getter method.

```suggestion
        "Session-Token": `${this._accessToken()}`,
```

lcaresia reviewed (changes requested):
One last change!

lcaresia reviewed (dismissed):
LGTM!

vunguyenhung commented:
Hello @feyzullah, I have tested this PR and there're some test cases failed or needed improvement. 

Please check the test report below for more information
[UpKeep_3823_2348.pdf](https://github.com/PipedreamHQ/pipedream/files/9508134/UpKeep_3823_2348.pdf)

feyzullah commented:
> 

Hi @vunguyenhung I have implemented your suggestions with one exception. It is 3rd suggestion on `Create Work Order`. Since categories can be added and removed for work orders and there is no method to fetch categories, I left `category` prop as it was. Can you please check the updates?

vunguyenhung commented:
Hi everyone, all test cases are passed! Ready for release! 

Test report
[UpKeep_3823_0036.pdf](https://github.com/PipedreamHQ/pipedream/files/9537647/UpKeep_3823_0036.pdf)

feyzullah commented:
/approve

jcortes reviewed (approved):
Approved!
